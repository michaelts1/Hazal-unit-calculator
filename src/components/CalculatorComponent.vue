<script lang="ts">
import type { DefaultUnits, Unit } from '../types'
import { defineComponent, PropType } from 'vue'
import InputAmount from './InputAmount.vue'
import { roundNum } from '../helpers'
import SelectUnit from './SelectUnit.vue'

export default defineComponent({
	components: {
		InputAmount,
		SelectUnit,
	},

	props: {
		units: {
			type: Array as PropType<Unit[]>,
			required: true,
		},
		defaultUnits: {
			type: Object as PropType<DefaultUnits>,
			required: true,
		},
	},

	data() {
		return {
			input: { unit: this.defaultUnits.input, value: 0 },
			output: { unit: this.defaultUnits.output, value: 0 },
		}
	},

	computed: {
		unitNames() { return this.units.map(unit => unit.name) },
	},

	watch: {
		units() {
			this.inputValueChanged(this.input.value)
		},
	},

	methods: {
		/** Converts measurements done using other units to metric */
		convertToMeters(amount: number, sourceUnit: string) {
			const unit = this.units.filter(({ name }) => name === sourceUnit)[0]
			return unit.value * amount
		},

		/** Converts metric measurements to other units */
		convertFromMeters(meters: number, targetUnit: string) {
			const unit = this.units.filter(({ name }) => name === targetUnit)[0]
			return meters / unit.value
		},

		/** Updates output with a new value */
		updateOutputValue(meters: number) {
			const outputValue = this.convertFromMeters(meters, this.output.unit)
			this.output.value = roundNum(outputValue)
		},

		/** Called after the input value has changed */
		inputValueChanged(newValue: number) {
			this.input.value = newValue
			const meters = this.convertToMeters(newValue, this.input.unit)
			this.updateOutputValue(meters)
		},

		/** Called after the input unit of measurement has changed */
		InputUnitChanged(newUnit: string) {
			this.input.unit = newUnit

			const meters = this.convertToMeters(this.input.value, newUnit)
			this.updateOutputValue(meters)
		},

		/** Called after the output unit of measurement has changed */
		OutputUnitChanged(newUnit: string) {
			const meters = this.convertToMeters(this.output.value, this.output.unit)
			this.output.unit = newUnit
			this.updateOutputValue(meters)
		},
	},
})
</script>

<template>
	<div class="wrapper-row calculator">
		<div class="wrapper-column">
			<InputAmount
				class="form-control form-control-sm"
				size="1"
				:value="input.value"
				@value-change="inputValueChanged"
			/>
			<SelectUnit
				class="form-select"
				:unit-names="unitNames"
				:selected-unit="input.unit"
				@value-change="newUnit => InputUnitChanged(newUnit)"
			/>
		</div>
		<div class="wrapper-column">
			=
		</div>
		<div class="wrapper-column">
			<InputAmount
				class="form-control form-control-sm"
				size="1"
				:disabled="true"
				:value="output.value"
			/>
			<SelectUnit
				class="form-select"
				:unit-names="unitNames"
				:selected-unit="output.unit"
				@value-change="newUnit => OutputUnitChanged(newUnit)"
			/>
		</div>
	</div>
</template>
