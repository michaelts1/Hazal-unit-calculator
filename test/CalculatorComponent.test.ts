import { describe, expect, it } from 'vitest'
import CalculatorComponent from '../src/components/CalculatorComponent.vue'
import { mount } from '@vue/test-utils'
import type { Unit } from '../src/types'

interface Field {
	unit: string
	value: number
}

describe('Component CalculatorComponent', () => {
	it('exists', () => {
		expect(CalculatorComponent).toBeTruthy()
	})

	const units: Unit[] = [
		{ hidden: false, name: 'אצבע', value: .005 },
		{ hidden: false, name: 'סנטימטר', value: .1 },
		{ hidden: true, name: 'אמה', value: 2 },
		{ hidden: false, name: 'טפח', value: 4 },
		{ hidden: false, name: 'זרת', value: 4000 },
	]
	const wrapper = mount(CalculatorComponent, {
		props: {
			units,
		},
	})

	it('responds to events emitted by SelectUnit', () => {
		const [ input, output ] = wrapper.findAllComponents({ name: 'SelectUnit' })

		expect((wrapper.vm.input as Field).unit).toBe('טפח') // Default
		input.vm.$emit('value-change', 'אצבע')
		expect((wrapper.vm.input as Field).unit).toBe('אצבע') // New value

		expect((wrapper.vm.output as Field).unit).toBe('סנטימטר') // Default
		output.vm.$emit('value-change', 'אמה')
		expect((wrapper.vm.output as Field).unit).toBe('אמה') // New value
	})

	it('responds to events emitted by InputAmount', () => {
		const input = wrapper.findComponent({ name: 'InputAmount' })

		expect((wrapper.vm.input as Field).value).toBe(0) // Default
		input.vm.$emit('value-change', 10)
		expect((wrapper.vm.input as Field).value).toBe(10) // New value
	})

	it('updates when the units prop changes', async () => {
		const oldValue = (wrapper.vm.output as Field).value

		const newUnits = JSON.parse(JSON.stringify(units))
		newUnits[0].value *= 2 // Change base unit value
		await wrapper.setProps({ units: newUnits })

		expect((wrapper.vm.output as Field).value).toBe(oldValue * 2)
	})
})
