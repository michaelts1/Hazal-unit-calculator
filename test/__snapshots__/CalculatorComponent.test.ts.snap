// Vitest Snapshot v1

exports[`Component CalculatorComponent > matches snapshot 1`] = `
"<div class=\\"wrapper-row\\" data-v-6764e49b=\\"\\">
  <div class=\\"wrapper-column\\" data-v-6764e49b=\\"\\"><select class=\\"centered\\" data-v-6764e49b=\\"\\">
      <option value=\\"אצבע\\">אצבע</option>
      <option value=\\"סנטימטר\\">סנטימטר</option>
      <option value=\\"אמה\\">אמה</option>
      <option value=\\"טפח\\">טפח</option>
      <option value=\\"זרת\\">זרת</option>
    </select><input class=\\"conversion-input\\" size=\\"12\\" data-v-519321c2=\\"\\" data-v-6764e49b=\\"\\"></div>
  <div class=\\"wrapper-column\\" data-v-6764e49b=\\"\\"> = </div>
  <div class=\\"wrapper-column\\" data-v-6764e49b=\\"\\"><select class=\\"centered\\" data-v-6764e49b=\\"\\">
      <option value=\\"אצבע\\">אצבע</option>
      <option value=\\"סנטימטר\\">סנטימטר</option>
      <option value=\\"אמה\\">אמה</option>
      <option value=\\"טפח\\">טפח</option>
      <option value=\\"זרת\\">זרת</option>
    </select><input class=\\"conversion-input\\" disabled=\\"\\" size=\\"12\\" data-v-519321c2=\\"\\" data-v-6764e49b=\\"\\"></div>
</div>"
`;
